pkg/storage/ingest/fetcher.go

```diff
@@ -0,0 +1,712 @@
+	return fetchResult{
+		ctx:            ctx,
+		fetchedBytes:   0,
+		FetchPartition: kgo.FetchPartition{Err: err},
+	}
+}
+
+type concurrentFetchers struct {
+	wg   sync.WaitGroup
+	done chan struct{}
+
+	client      *kgo.Client
+	logger      log.Logger
+	partitionID int32
+	topicID     [16]byte
+	topicName   string
+	metrics     *readerMetrics
+	tracer      *kotel.Tracer
+
+	minBytesWaitTime time.Duration
+
+	orderedFetches     chan fetchResult
+	lastReturnedRecord int64
+	startOffsets       *genericOffsetReader[int64]
+
+	// trackCompressedBytes controls whether to calculate MaxBytes for fetch requests based on previous responses' compressed or uncompressed bytes.
+	trackCompressedBytes bool
+}
+
+func (r *concurrentFetchers) Stop() {
+	close(r.done)
+
+	r.wg.Wait()
+	level.Info(r.logger).Log("msg", "stopped concurrent fetchers", "last_returned_record", r.lastReturnedRecord)
+}
+
+// newConcurrentFetchers creates a new concurrentFetchers. startOffset can be kafkaOffsetStart, kafkaOffsetEnd or a specific offset.
+func newConcurrentFetchers(
+	ctx context.Context,
+	client *kgo.Client,
+	logger log.Logger,
+	topic string,
+	partition int32,
+	startOffset int64,
+	concurrency int,
+	recordsPerFetch int,
+	trackCompressedBytes bool,
+	minBytesWaitTime time.Duration,
+	offsetReader *partitionOffsetClient,
+	startOffsetsReader *genericOffsetReader[int64],
+	metrics *readerMetrics,
+) (*concurrentFetchers, error) {
+
+	var err error
+	switch startOffset {
+	case kafkaOffsetStart:
+		startOffset, err = offsetReader.FetchPartitionStartOffset(ctx, partition)
+	case kafkaOffsetEnd:
+		startOffset, err = offsetReader.FetchPartitionLastProducedOffset(ctx, partition)
+		// End (-1) means "ignore all existing records". FetchPartitionLastProducedOffset returns the offset of an existing record.
+		// We need to start from the next one, which is still not produced.
+		startOffset++
+	}
+	if err != nil {
+		return nil, fmt.Errorf("resolving offset to start consuming from: %w", err)
+	}
+	f := &concurrentFetchers{
+		client:               client,
+		logger:               logger,
+		topicName:            topic,
+		partitionID:          partition,
+		metrics:              metrics,
+		minBytesWaitTime:     minBytesWaitTime,
+		lastReturnedRecord:   startOffset - 1,
+		startOffsets:         startOffsetsReader,
+		trackCompressedBytes: trackCompressedBytes,
+		tracer:               recordsTracer(),
+		orderedFetches:       make(chan fetchResult),
+		done:                 make(chan struct{}),
+	}
+
+	topics, err := kadm.NewClient(client).ListTopics(ctx, topic)
+	if err != nil {
+		return nil, fmt.Errorf("failed to find topic ID: %w", err)
+	}
+	if !topics.Has(topic) {
+		return nil, fmt.Errorf("failed to find topic ID: topic not found")
+	}
+	f.topicID = topics[topic].ID
+
+	f.wg.Add(1)
+	go f.start(ctx, startOffset, concurrency, recordsPerFetch)
+
+	return f, nil
+}
+
+func (r *concurrentFetchers) Update(ctx context.Context, concurrency, records int) {
+	r.Stop()
```

Comment:
nit: Can you add a task to not update concurrency if we are already at that one? With the change we made as we approach the desired lag, there's a chance that we're already at the specified concurrency and we've stopped and start the workers for no reason.